experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts

```diff
@@ -497,3 +514,76 @@ export function createListProxy<TTypes extends AllowedTypes>(
 		},
 	});
 }
+
+// #region Create dispatch map for maps
+
+const mapStaticDispatchMap: PropertyDescriptorMap = {
+	[Symbol.iterator]: {
+		value(this: SharedTreeMap<MapSchema>) {
+			const node = getMapNode(this);
+			return node[Symbol.iterator]();
+		},
+		writable: false,
+		enumerable: false,
+		configurable: false,
+	},
+	get: {
+		value(key: string): unknown {
+			const mapNode = getMapNode(this);
+			const field = mapNode.getBoxed(key);
+			return getProxyForField(field);
+		},
+	},
+	entries: {
+		value(): IterableIterator<[string, unknown]> {
+			const mapNode = getMapNode(this);
+			return mapNode.entries();
+		},
+	},
+	// TODO: implement the rest of the Map interface here
+};
+
+const mapPrototype = Object.create(Object.prototype, mapStaticDispatchMap);
+
+// #endregion
+
+const getMapNode = <TSchema extends MapSchema>(target: object): MapNode<TSchema> => {
+	return getTreeNode(target) as MapNode<TSchema>;
+};
+
+function createMapProxy<TSchema extends MapSchema>(treeNode: TreeNode): SharedTreeMap<TSchema> {
+	// Create a 'dispatch' object that this Proxy forwards to instead of the proxy target.
+	// Own properties on the dispatch object are surfaced as own properties of the proxy.
+	// (e.g., 'size', which is defined below).
+	const dispatch = Object.create(mapPrototype, {
+		size: {
+			get(this: object) {
+				return getMapNode(this).size;
+			},
+		},
+	});
+
+	setTreeNode(dispatch, treeNode);
+
+	// TODO: Although the target is an object literal, it's still worthwhile to try experimenting with
+	// a dispatch object to see if it improves performance.
+	return new Proxy<SharedTreeMap<TSchema>>(new Map<string, ProxyNode<TSchema>>(), {
+		get: (target, key, receiver): unknown => {
+			return Reflect.get(dispatch, key);
+		},
+		getOwnPropertyDescriptor: (target, key): PropertyDescriptor | undefined => {
+			return Reflect.getOwnPropertyDescriptor(dispatch, key);
+		},
+		has: (target, key) => {
+			return Reflect.has(dispatch, key);
+		},
+		set: (target, key, newValue): boolean => {
+			// For MVP, we otherwise disallow setting properties (mutation is only available via the map mutation APIs).
+			return false;
+		},
+		ownKeys: (target) => {
+			// TODO: double check that this is all we require here
+			return ["size"];
```

Comment:
Interesting.  On Map, all properties (inculding 'size') are inherited via the prototype:

```
> Reflect.ownKeys(new Map([[0,1]]))
[]
```

So 'ownKeys' should return [], getOwnPropertyDescriptor always returns undefined, etc.

Reply:
Ah, I see. And thanks for the tip on how to explicitly verify! I'll get this updated and add some comments!